coral-service/src/main/java/com/linkedin/coral/coralservice/utils/VisualizationUtils.java

```diff
@@ -0,0 +1,128 @@
+import com.linkedin.coral.incremental.RelNodeIncrementalTransformer;
+import com.linkedin.coral.transformers.CoralRelToSqlNodeConverter;
+import com.linkedin.coral.trino.trino2rel.TrinoToRelConverter;
+import com.linkedin.coral.vis.VisualizationUtil;
+
+import static com.linkedin.coral.coralservice.utils.CoralProvider.*;
+
+
+public class VisualizationUtils {
+
+  public static File getImageDir() {
+    return new File(System.getProperty("java.io.tmpdir") + "/images" + UUID.randomUUID());
+  }
+
+  public boolean isValidFromLanguage(String fromLanguage) {
+    return fromLanguage.equalsIgnoreCase("trino") || fromLanguage.equalsIgnoreCase("hive");
+  }
+
+  public ArrayList<UUID> generateIRVisualizations(String query, String fromLanguage, File imageDir,
+      RewriteType rewriteType) {
+    ArrayList<UUID> imageIDList = new ArrayList<>();
+
+    // Always generate the pre/no rewrite images first
+    RelNodeAndID relNodeAndID = generateRelNodeVisualization(getRelNode(query, fromLanguage), imageDir);
+    imageIDList.add(relNodeAndID.id);
+
+    SqlNodeAndID sqlNodeAndID = generateSqlNodeVisualization(getSqlNode(query, fromLanguage), imageDir);
+    imageIDList.add(sqlNodeAndID.id);
+
+    // Generate rewritten IR images if requested, otherwise, simply return the non rewritten image ids
+    RelNode preRewriteRelNode = relNodeAndID.relNode;
+    RelNode postRewriteRelNode;
+
+    if (rewriteType != RewriteType.NONE && rewriteType != null) {
+      switch (rewriteType) {
+        case INCREMENTAL:
+          postRewriteRelNode = RelNodeIncrementalTransformer.convertRelIncremental(preRewriteRelNode);
+          break;
+        case DATAMASKING:
+        default:
+          return imageIDList;
+      }
+      RelNodeAndID postRewroteRelNodeAndID = generateRelNodeVisualization(postRewriteRelNode, imageDir);
+      imageIDList.add(postRewroteRelNodeAndID.id);
+
+      SqlNode postRewriteSqlNode = new CoralRelToSqlNodeConverter().convert(postRewriteRelNode);
+      SqlNodeAndID postRewriteSqlNodeAndID = generateSqlNodeVisualization(postRewriteSqlNode, imageDir);
+      imageIDList.add(postRewriteSqlNodeAndID.id);
+    }
+
+    return imageIDList;
+  }
+
+  private SqlNodeAndID generateSqlNodeVisualization(SqlNode sqlNode, File imageDir) {
+    // Generate graphviz svg using sqlNode
+    VisualizationUtil visualizationUtil = VisualizationUtil.create(imageDir);
+    UUID sqlNodeId = UUID.randomUUID();
+    visualizationUtil.visualizeSqlNodeToFile(sqlNode, "/" + sqlNodeId + ".svg");
+
+    return new SqlNodeAndID(sqlNodeId, sqlNode);
+  }
+
+  private RelNodeAndID generateRelNodeVisualization(RelNode relNode, File imageDir) {
+    // Generate graphviz svg using relNode
+    VisualizationUtil visualizationUtil = VisualizationUtil.create(imageDir);
+    UUID relNodeID = UUID.randomUUID();
+    visualizationUtil.visualizeRelNodeToFile(relNode, "/" + relNodeID + ".svg");
+
+    return new RelNodeAndID(relNodeID, relNode);
+  }
```

Comment:
You are passing in the relNode and returning it back, and creating a new class `RelNodeAndID` to accommodate that. Does not seem like this is necessary.

Reply:
Fixed!